From 31f3ffab952f0a75066eaf7bb56359e2ea974d05 Mon Sep 17 00:00:00 2001
From: Vladimir Kostyukov <vladimir.kostyukov@intel.com>
Date: Thu, 27 Jun 2013 17:05:06 +0700
Subject: Dalvik: A correct assert statement for determineVirtualRegisterState

BZ: 119535

The assert is failed since compiler doesn't expect to find a constant
in association table, which is, actually, possible. This patch includes
additional test for the case when virtual register is a constant, which
doesn't associated with a physical register.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: I06cfa6a180f56dedf5c85c601d50c8f320ed6ad3
Orig-MCG-Change-Id: I5d99bd11ec1e5b7aa8e77f0a78e1f6c8a93c0b58
Signed-off-by: Vladimir Kostyukov <vladimir.kostyukov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/RegisterizationBE.cpp |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/codegen/x86/RegisterizationBE.cpp b/vm/compiler/codegen/x86/RegisterizationBE.cpp
index 689b9f0..1cf4d0f 100644
--- a/vm/compiler/codegen/x86/RegisterizationBE.cpp
+++ b/vm/compiler/codegen/x86/RegisterizationBE.cpp
@@ -503,11 +503,13 @@ static bool determineVirtualRegisterState (const AssociationTable &associations,
         {
             const CompileTableEntry &compileEntry = assocIter->second;
 
+            bool inPhysicalReg = compileEntry.inPhysicalRegister ();
             bool inGP = compileEntry.inGeneralPurposeRegister ();
             bool inXMM = compileEntry.inXMMRegister ();
 
             //In order to have saved it, it must have been in either GP or XMM
-            assert ((inGP || inXMM) == true);
+            //It also can be in a contant, which doesn't associated with physical reg
+            assert (inPhysicalReg == false || (inGP || inXMM) == true);
 
             if (inGP == true)
             {
@@ -530,7 +532,7 @@ static bool determineVirtualRegisterState (const AssociationTable &associations,
                     continue;
                 }
             }
-            else if (compileEntry.inPhysicalRegister ())
+            else if (inPhysicalReg == true)
             {
                 ALOGI ("JIT_INFO: We failed to satisfy BB associations because we found a VR that "
                         "is in physical register but not in GP or XMM.");
-- 
1.7.4.1

